### ISSUE TITLE
CUDA 13.3 ptxas SIGSEGV compiling sm120 sparse MQA logits (MXFP4, non-WS path)

### ISSUE THREAD

**Issue by lucifer1004** [lucifer1004/DeepGEMM-sm120#2]
## Summary

`sm120_fp8_fp4_sparse_mqa_logits_kernel` (vendored `impls/sm120_fp8_fp4_sparse_mqa_logits.cuh`) fails to JIT-compile under **CUDA 13.3** (13.3.73): the nvcc compile subprocess dies with SIGSEGV (exit 139). The same sources compile and pass all tests under **CUDA 13.2** (13.2.86), the documented validation toolchain. Root cause is a ptxas codegen crash in CUDA 13.3 — **not a kernel source bug**.

## Reproduction

On any SM120 GPU with the nv_dev or vllm-project fork (or any DeepJIT consumer of this kernel), using a CUDA 13.3 toolkit:

```
pytest tests/test_sm120_attention.py::test_sm120_sparse_contiguous   # contiguous
pytest tests/test_sm120_attention.py::test_sm120_sparse_paged        # paged
```

Both fail with `exit code 139` from the DeepJIT nvcc subprocess. (Reproduced on RTX PRO 6000 Blackwell, sm_120a.)

## Characterization

- **Crashing component**: `ptxas` only. `nvcc -ptx` (cicc frontend) produces valid PTX under 13.3; feeding that same PTX to 13.3 ptxas crashes.
- **Trigger**: ptxas `-O3` (the default) **and** any `--register-usage-level` value (6–10 all crash). DeepJIT's default options include `--ptxas-options=--register-usage-level=10`, so stock builds hit this.
  - `-O2` + register-usage-level=10 → OK
  - `-O3` with no register-usage-level → OK
- **Affected instantiations** (probed via captured JIT sources):
  - MXFP4 (`is_fp4=true`), **non-warp-specialized**, contiguous (P1/P2, pipelined or not) → crash
  - MXFP4 paged (per the original test failure) → crash
  - FP8 (`is_fp4=false`) non-WS → OK
  - MXFP4 warp-specialized (`work_partitions=4`) → OK
  - All other SM120 kernels (dense/paged MQA, full GEMM suite, sparse MQA metadata; 100 tests) compile and pass under 13.3
- **Arch**: crashes for both `sm_120a` and `sm_121f` targets — SM121 is affected identically.

## Workarounds (validated manually, not implemented)

Either of these ptxas-side adjustments avoids the crash for the failing instantiation:

- `nvcc ... --ptxas-options=--register-usage-level=10 --ptxas-options=-O2` (separate `--ptxas-options` args compose), or
- drop `--register-usage-level` entirely.

The `-O2` variant produces **identical register allocation to the validated 13.2 build** for the probed instantiation (REG:101, STACK:24, SHARED:1024 in both). A scoped implementation would be a per-kernel `CompilerOptions::extra_nvcc_flags` override in each fork's host launcher (`csrc/jit_kernels/impls/sm120_fp8_fp4_sparse_mqa_logits.hpp`), gated on nvcc >= 13.3 — the canonical repo is headers-only, so the flag cannot live here.

## Notes

- CUDA 13.2 remains the validated toolchain (see README); SM121 hardware is otherwise unvalidated.
- A ptxas crash is a compiler bug; if a minimal LLVM-IR/PTX reproducer is needed for an NVIDIA report, the failing PTX and kernel.cu can be extracted from DeepJIT's cache tmp dir (or via an nvcc wrapper that copies the `.cu` before invoking the real compiler).

**Comment by lucifer1004** [comment]
Fixed on the nv_dev-lineage fork: deepseek-ai/DeepGEMM#447 commit `f7bea76` adds a toolchain-gated per-kernel override (nvcc >= 13.3 → ptxas `-O2` for the non-WS instantiations, via DeepJIT `CompilerOptions::extra_nvcc_flags`). Validated on sm_120a: sparse MQA suite 9/9 under CUDA 13.3.73 with the gate active, 9/9 under 13.2.86 with the gate inactive. The vllm-project fork (#10) needs no change — it vendors the device headers but has no SM120 sparse-MQA host glue, so nothing instantiates this kernel there. Note the arch probe: the crash reproduces for `sm_121f` targets too, so SM121 consumers would hit the same issue.

**Comment by lucifer1004** [comment]
**CUDA 13.4 does not fix this.** Tested 13.4.59: end-to-end nvcc compile of the failing instantiation still exits 139, and ptxas 13.4 still SIGSEGVs on the same PTX. The fork-side gate in deepseek-ai/DeepGEMM#447 (`f7bea76`) uses nvcc `>= 13.3`, so 13.4 is covered.

A minimal self-contained reproducer is now available (single 42 KB PTX file, no headers needed):
```
ptxas --register-usage-level=10 kernel.ptx -o kernel.cubin --gpu-name=sm_120f
```
Same PTX: ptxas 13.2.86 OK · 13.3.73 SIGSEGV · 13.4.59 SIGSEGV. (PTX generated by the 13.2 frontend, so 13.2 serves as a positive control.) Ready to hand to NVIDIA contacts; tracked locally at the inferlab scratchpad `dsv41-sm120/cuda133-repro/ptxas-sigsegv-repro/`.
